### PR TITLE
Duplicate Ways Check

### PR DESCRIPTION
**Pull request by savannahostrowski**
This pull request covers the logic, unit tests, documentation, and configuration for the Duplicate Ways check. 

The Duplicate Ways check flags Ways in OpenStreetMap which are either partially or completely duplicated.

### CODE REVIEW

**Comment by danielduhh** [comment]
Missing DuplicateWays in configuration

**Comment by savannahostrowski** [comment]
@danielduhh Do we actually want to add DuplicateWays to the configuration? There's nothing to configure for this check. I suppose I could still add it for challenge information.

**Comment by danielduhh** [comment]
Yeah, configuration for this would be for MapRoulette. It's nice to have present because I can set "enabled": true/false to specify whether I want this check to run through Spark when running as standalone application

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * @author savannahostrowski
+ */
+public class DuplicateWaysCheck extends BaseCheck
+{
+
+    // You can use serialver to regenerate the serial UID.
+    private static final long serialVersionUID = 1L;
+
+    public static final String DUPLICATE_EDGE_INSTRUCTIONS = "This way, {0,number,#}, "
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: typo?

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                continue;
+            }
+
+            // Check if the Segment is in globalSegments
+            if (globalSegments.containsKey(segment))
+            {
+                // add identifier to the list of identifiers with that segment
+                globalSegments.get(segment).add(identifier);
+
+                if (globalSegments.get(segment).size() > 1
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Since we already know the segment exists in `globalSegments`, and are adding the current `segment` on line 88, can't we assume that globalSegments.get(segment).size() will always be above 1? I think we can get away with..
```java
            if (globalSegments.containsKey(segment))
            {
                if (!this.isFlagged(edge.getMasterEdgeIdentifier()))
                {
                    this.markAsFlagged(edge.getMasterEdgeIdentifier());
                    return Optional.of(this.createFlag(edge, this.getLocalizedInstruction(0,
                            edge.getOsmIdentifier())));
                }
            }
```

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    public static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+
+    public static final int ZERO_LENGTH = 0;
+    public static final String AREA_KEY = "area";
+
+
+    // a map of segments and list of Edge identifiers
+    private final Map<Segment, Set<Long>> globalSegments = new HashMap<>();
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Whats the advantage of storing a `HashMap`? Couldn't we just create a `Set<Segment>` & add use .contains() to check if segment exists in `globalSegments`?

File: docs/checks/duplicateWaysCheck.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                    identifiers.add(identifier);
+                    globalSegments.put(segment, identifiers);
+                }
+            }
+    
+            return Optional.empty();
+        }
+```
+
+Within the check body, we get all Segments in each Edge, store each unique Segment as a key in a
```
Comment (danielduhh) [inline docs/checks/duplicateWaysCheck.md]: extra credit: I think its worth mentioning what you're doing on line 64

File: data/README.md [hunk 5]
```diff
@@ -1 +0,0 @@
-Place Atlas or OSM pbf files for processing in this directory
```
[the remark this answers is not in the document]
Comment (savannahostrowski) [reply data/README.md]: Not sure, I'll fix it.

File: .gitignore [hunk 6]
```diff
@@ -1,21 +0,0 @@
-*/build/
```
[the remark this answers is not in the document]
Comment (savannahostrowski) [reply .gitignore]: I .gitignored it for development purposes. I'll re-add it.

File: docs/checks/duplicateWaysCheck.md (hunk 4 above)
Comment (savannahostrowski) [reply docs/checks/duplicateWaysCheck.md]: Alrighty!

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java (hunk 3 above)
Comment (savannahostrowski) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: The problem lies with the Segment objects containing no information about the Edge it came from. We can go from an Edge into its Segments but not from its Segments back to the Edge. In using a HashMap, we can store a Segment keep the mapping of the Edge it came from.

But yeah, if that functionality did exist, you're right. We could just use a HashSet and still retain time complexity.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java (hunk 2 above)
Comment (savannahostrowski) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Yep, you're right! Nice one!

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            // Check if the Segment is in globalSegments
+            if (globalSegments.containsKey(segment))
+            {
+                // add identifier to the list of identifiers with that segment
+                globalSegments.get(segment).add(identifier);
+
+                if (globalSegments.get(segment).size() > 1
+                        && !this.isFlagged(edge.getMasterEdgeIdentifier()))
+                {
+                    this.markAsFlagged(edge.getMasterEdgeIdentifier());
```
[the remark this answers is not in the document]
Comment (savannahostrowski) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Nice. I'll change them to match the master edge.

File: docs/checks/duplicateWaysCheck.md [hunk 8]
```diff
@@ -91,11 +90,12 @@ statements to validate whether we do in fact want to flag the feature for inspec
         }
 ```
 
-Within the check body, we get all Segments in each Edge, store each unique Segment as a key in a 
-HashMap, and store each Edge which contains that Segment as the value. If that Segment already exists
-in the HashMap (so the list of Edges is greater than 1) and the Edge identifier has not already been
-flagged, we flag the Edge. If that Segment does not already contain that key, we add the Segment,
-Edge identifier key value pair into the HashMap.
+Within the check body, we first check that the Segment is greater than zero meters in length to
+ensure that we are not just looking at a duplicated node. Next,we get all Segments in each Edge,
```
Comment (danielduhh) [inline docs/checks/duplicateWaysCheck.md]: nit: missing space before comma ",we get all Segments"

**Comment by savannahostrowski** [comment]
@danielduhh Addressed your comments and squashed some commits. Let me know if there are any problems.
